source/particle/property/viscoplastic_strain_invariants.cc

```diff
@@ -107,11 +109,36 @@ namespace aspect
             material_inputs.composition[0][i] = solution[this->introspection().component_indices.compositional_fields[i]];
           }
 
-        // Find out plastic yielding by calling function in material model.
+        /* Find out plastic yielding by calling function in material model. Here we also deal with phases
+        in the material model, which may affect how the viscosity is calculated, thus affect the result.
+        For this an approach that is similar to ViscoPlastic::evalute is implememnted. Although the private
+        member of that class cannot be accessed here, public functions accessed instead.*/
         const MaterialModel::ViscoPlastic<dim> &viscoplastic
           = Plugins::get_plugin_as_type<const MaterialModel::ViscoPlastic<dim>>(this->get_material_model());
 
-        const bool plastic_yielding = viscoplastic.is_yielding(material_inputs);
+        const double gravity_norm = this->get_gravity_model().gravity_vector(material_inputs.position[0]).norm();
+
+        double reference_density;
+        if (this->get_adiabatic_conditions().is_initialized())
+          {
+            reference_density = this->get_adiabatic_conditions().density(material_inputs.position[0]);
+          }
+        else
+          {
+            MaterialModel::EquationOfStateOutputs<dim> eos_outputs_all_phases (this->n_compositional_fields()+1+viscoplastic.get_phase_function_n_phase_transitions());
+            viscoplastic.equation_of_state_evaluate(material_inputs, 0, eos_outputs_all_phases);
+            reference_density = eos_outputs_all_phases.densities[0];
+          }
+
+        MaterialModel::MaterialUtilities::PhaseFunctionInputs<dim> phase_inputs(material_inputs.temperature[0],
+                                                                                material_inputs.pressure[0],
+                                                                                this->get_geometry_model().depth(material_inputs.position[0]),
+                                                                                gravity_norm*reference_density,
+                                                                                numbers::invalid_unsigned_int);
+
+        std::vector<double> phase_function_values = viscoplastic.phase_function_compute_value(phase_inputs);
```

Comment:
As just discussed: If you move this code into `ViscoPlastic::is_yielding` you do not need the new functions inside `ViscoPlastic` and can save some code.

Reply:
Thanks, I have modified the codes according to your suggestions. Now please take a look again and see what to do next.